src/coinsviewcacheasync.h

```diff
@@ -0,0 +1,242 @@
+    std::optional<uint32_t> GetInputIndex(const COutPoint& outpoint) const noexcept
+    {
+        // This assumes ConnectBlock accesses all inputs in the same order as they are added to m_inputs
+        // in StartFetching. Some outpoints are not accessed because they are created by the block, so we scan until we
+        // come across the requested input. We advance the tail since the input will be cached and not accessed through
+        // this method again.
+        for (const auto i : std::views::iota(m_input_tail, m_inputs.size())) [[likely]] {
+            if (m_inputs[i].outpoint == outpoint) {
+                m_input_tail = i + 1;
+                return i;
+            }
+        }
+        return std::nullopt;
+    }
+
+    CCoinsMap::iterator FetchCoin(const COutPoint& outpoint) const override
+    {
+        auto [ret, inserted]{cacheCoins.try_emplace(outpoint)};
+        if (!inserted) return ret;
+
+        if (const auto i{GetInputIndex(outpoint)}) [[likely]] {
+            auto& input{m_inputs[*i]};
+            // Check if the coin is ready to be read. We need to acquire to match the worker thread's release.
+            while (!input.ready.test(std::memory_order_acquire)) {
+                // Work instead of waiting if the coin is not ready
+                if (!ProcessInputInBackground()) {
+                    // No more work, just wait
+                    input.ready.wait(/*old=*/false, std::memory_order_acquire);
+                    break;
+                }
+            }
+            if (input.coin) [[likely]] ret->second.coin = std::move(*input.coin);
+        }
+
+        if (ret->second.coin.IsSpent()) [[unlikely]] {
+            // We will only get in here for BIP30 checks, shorttxid collisions, or a block with missing or spent inputs.
+            // TODO: Remove spent checks once we no longer return spent coins in coinscache_sim CoinsViewBottom.
+            if (auto coin{FetchCoinWithoutMutating(outpoint)}; coin && !coin->IsSpent()) {
+                ret->second.coin = std::move(*coin);
+            } else {
+                cacheCoins.erase(ret);
+                return cacheCoins.end();
+            }
+        }
+
+        cachedCoinsUsage += ret->second.coin.DynamicMemoryUsage();
+        return ret;
+    }
+
+    std::vector<std::thread> m_worker_threads{};
+    std::barrier<> m_barrier;
+
+    //! Stop all worker threads.
+    void StopFetching() noexcept
+    {
+        if (m_inputs.empty()) return;
+        // Skip fetching the rest of the inputs by moving the head to the end.
+        m_input_head.store(m_inputs.size(), std::memory_order_relaxed);
+        // Wait for all threads to stop.
+        m_barrier.arrive_and_wait();
+        m_inputs.clear();
+        m_input_head.store(0, std::memory_order_relaxed);
+        m_input_tail = 0;
+        m_txids.clear();
+    }
+
+public:
+    //! Fetch all block inputs.
+    void StartFetching(const CBlock& block) noexcept
```

Comment:
Is this idempotent, are we sure all state is reset after the previous block? Could we call a reset here or an assert to make sure we're not accidentally inheriting anything from a previous (failed?) fetch?